### PR TITLE
Open tracking url in a new tab and add the link to the order preview

### LINKED ISSUE #23401: [BO] Carrier Tracking URLs in PS 1.7.7.0+
**Describe the Bug:**
After BO Order Admin has been migrated to Symphony, a few issues related to the Carrier tracking numbers have appeared

1. The Carrier tracking link no longer opens in a new browser tab ("target="_blank") in BO **Order Details** page, but in self-page.
2. The Tracking number is not a link on the **Order List** page (dropdown)

**Expected Behavior:**
1. Carrier tracking links in BO **Order Details** page should open in new browser tab ("target="_blank")
ex on PS 1.6.1.24 :
Capture d’écran 2021-02-26 à 17 51 54

2. The Tracking number on the **Order List** page (dropdown) should also be the same url as well, and also open in new browser tab ("target="_blank") as well.
Capture d’écran 2021-02-26 à 17 51 14

**Steps to reproduce:**
- Add a Carrier tracking number to an order (from a carrier with a pre-defined tracking URL included an @ to represent the tracking number).
- Click on the Carrier tracking url in BO Order page (will open in same window/tab).

**Additional Information:**
Prestashop version: 1.7.7.2
PHP Version: 7.3

### PR DESCRIPTION
**Pull request by atomiix**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.7.x
| Description?      | Tracking url links should open in a new tab. Also, there was no link in the order preview on the order list page. This PR fixes both issues.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23401
| How to test?      | Please see #23401
| Possible impacts? | 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23819)

### CODE REVIEW

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Could you modify unit tests and add behat tests, please ?

File: tests/Integration/Behaviour/Features/Context/CarrierFeatureContext.php [hunk 1]
```diff
@@ -258,6 +258,17 @@ public function setCarrierShipsToAllGroups($carrierName)
         $carrier->setGroups($groupIds);
     }
 
+    /**
+     * @Given /^carrier "(.+)" uses "(.+)" as tracking url$/
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/CarrierFeatureContext.php]:
```suggestion
     * @Given /^the carrier "(.+)" uses "(.+)" as tracking url$/
```

File: tests/Integration/Behaviour/Features/Context/CarrierFeatureContext.php [hunk 2]
```diff
@@ -258,6 +258,17 @@ public function setCarrierShipsToAllGroups($carrierName)
...
+     */
+    public function setCarrierTrackingUrl($carrierName, $url)
```
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/CarrierFeatureContext.php]:
```suggestion
    public function setCarrierTrackingUrl(string $carrierName, string $url): void
```

File: src/Adapter/Order/QueryHandler/GetOrderPreviewHandler.php [hunk 3]
```diff
@@ -151,10 +151,12 @@ private function getShippingDetails(Order $order): OrderPreviewShippingDetails
         $state = new State($address->id_state);
 
         $carrierName = null;
+        $trackingUrl = null;
```
Comment (Progi1984) [inline src/Adapter/Order/QueryHandler/GetOrderPreviewHandler.php]:
```suggestion
        $carrierName = $trackingUrl = null;
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/shipping.html.twig [hunk 4]
```diff
@@ -59,7 +59,7 @@
           <td>
             {% if carrier.trackingNumber %}
               {% if carrier.trackingUrl %}
-                <a href="{{ carrier.trackingUrl }}">{{ carrier.trackingNumber }}</a>
+                <a href="{{ carrier.trackingUrl }}" target="_blank">{{ carrier.trackingNumber }}</a>
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/View/shipping.html.twig]:
```suggestion
                <a href="{{ carrier.trackingUrl }}" target="_blank" rel="noopener noreferrer nofollow">{{ carrier.trackingNumber }}</a>
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/preview.html.twig [hunk 5]
```diff
@@ -47,7 +47,11 @@
               <p class="mb-0">
                 <strong>{{ 'Tracking number'|trans({}, 'Admin.Shipping.Feature') }}:</strong>
                 {% if not orderPreview.isVirtual and orderPreview.shippingDetails.trackingNumber is not empty %}
-                  {{ orderPreview.shippingDetails.trackingNumber }}
+                  {% if orderPreview.shippingDetails.trackingUrl %}
+                    <a href="{{ orderPreview.shippingDetails.trackingUrl }}" target="_blank">{{ orderPreview.shippingDetails.trackingNumber }}</a>
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/preview.html.twig]:
```suggestion
                    <a href="{{ orderPreview.shippingDetails.trackingUrl }}" target="_blank" rel="noopener noreferrer nofollow">{{ orderPreview.shippingDetails.trackingNumber }}</a>
```

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Sorry to block this PR, it's a great job, you even add test without waiting for @Progi1984 's comment :trollface: 
But I'm afraid there are security issue, you always have to set ` rel="noopener noreferrer nofollow"`when using `target="_blank"` to another website.

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @atomiix,

It is Ok :heavy_check_mark: 
In the FO, I have the same behavior as PS1769.
I updated the tracking number from the BO and from API => ok :heavy_check_mark: 
https://drive.google.com/file/d/18icC7i_KO1aMuEhejHIN_8fFMftlxus0/view?usp=sharing

Thank you!

**Comment by Progi1984** [comment]
Thanks @atomiix & @khouloudbelguith

**Comment by kmorgen** [comment]
Just updated a dev site to PS 1.7.7.4, and noticed the tracking url on the **Order List** page (dropdown) is incorrect.
It does not include the actual barcode number in the url like on the **Order details** page.

(The @ is not replaced with the actual tracking number in the Order list link, so it becomes pretty unusable here)

Did nobody do any QA?

**Comment by matks** [comment]
Hello @kmorgen 

Please remember this project follows a [Code of Conduct](https://github.com/PrestaShop/PrestaShop/blob/17196f515d27b02d9d8318f7da95e3907bde9d5d/CODE_OF_CONDUCT.md) that recommends

> Using welcoming and inclusive language
Being respectful of differing viewpoints and experiences
Gracefully accepting constructive criticism
Focusing on what is best for the community
Showing empathy towards other community members

<hr>

> Did nobody do any QA?

If I am not wrong, this is [sarcasm](https://en.wikipedia.org/wiki/Sarcasm) and sarcasm is against our CoC.

This PR has been marked "QA ✔️ " as it was tested by someone. There is even a video record of what has been tested. If you think the topic can be improved, feel free to submit a Pull Request.

**Comment by kmorgen** [comment]
> If I am not wrong, this is [sarcasm](https://en.wikipedia.org/wiki/Sarcasm) and sarcasm is against our CoC.
> 
> This PR has been marked "QA ✔️ " as it was tested by someone. There is even a video record of what has been tested. If you think the topic can be improved, feel free to submit a Pull Request.

It was not meant as sarcastic, but actually as a question, as I don't understand the QA passing process.

I was one of the original reporters of this issue, and it always takes a great effort to report/document new issues and to reproduce them. But when being fixed, the initial reporters seems not actively included as part of the QA process prior to release.

As a result of this many issues seem to need re-work for several updates over months/years, and often new issues are being introduced in the process unfortunately.

It litterly took med 10 seconds to verify the Order List page url was not working properly after updating to 1.7.7.4, as I already knew about the issue. I really think you should consider to include the initial bug/feature reporters as a mandatory part of the QA process somehow. As they are the ones dealing with the issues, they can also verify if the various fixes works as intended - even if they might not be coders.

(How/where do I see this QA process and video records ect, When clicking the "QA ✔️ " image, I just end up here: https://github.com/PrestaShop/PrestaShop/labels/QA%20%E2%9C%94%EF%B8%8F 
Is i possible to be part of the QA process prior to releases, even if not a skilled coder?

**Comment by kmorgen** [comment]
Steps to reproduce the fix is not working as intended;

1. Add a Carrier tracking number to an order.
  (from a carrier with a pre-defined tracking URL containing the @ representing the tracking number part).
  
2. Compare the Carrier tracking url in **BO Order** page vs, the **Order List** page.
(the url on the Order List does not contain the tracking number, but only the raw pre-set url from Carrier settings. It should be identical to the Order details page)

**Comment by matks** [comment]
> It was not meant as sarcastic, but actually as a question, as I don't understand the QA passing process.
> 
> I was one of the original reporters of this issue, and it always takes a great effort to report/document new issues and to reproduce them. But when being fixed, the initial reporters seems not actively included as part of the QA process prior to release.
> 
> As a result of this many issues seem to need re-work for several updates over months/years, and often new issues are being introduced in the process unfortunately.
> 
> It litterly took med 10 seconds to verify the Order List page url was not working properly after updating to 1.7.7.4, as I already knew about the issue. I really think you should consider to include the initial bug/feature reporters as a mandatory part of the QA process somehow. As they are the ones dealing with the issues, they can also verify if the various fixes works as intended - even if they might not be coders.
> 
> (How/where do I see this QA process and video records ect, When clicking the "QA ✔️ " image, I just end up here: https://github.com/PrestaShop/PrestaShop/labels/QA%20%E2%9C%94%EF%B8%8F
> Is i possible to be part of the QA process prior to releases, even if not a skilled coder?

Hello @kmorgen thank you for your answer it is very constructive.

I said "If I am not wrong, this is sarcasm and sarcasm is against our CoC." . It looks like I was wrong 😊  and I knew it could happen, which is why I made sure to include "If I am not wrong" at the beginning of the sentence.

> >Is i possible to be part of the QA process prior to releases, even if not a skilled coder

That is actually a very good idea 👍  although I am not sure a lot of people who report bugs will go as far as testing the bugfix. But we can try.

Thank you for your very good idea I will reach out to the team to see if we can improve our QA process to include the bug reporter.

**Pull request merged**